### PR TITLE
Kier review

### PR DESCRIPTION
**Pull request by eanders-ms**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,13 +1,91 @@
...
+git clone https://github.com/KierPalin/pxt-microbit 
+cd pxt-microbit
+```
+3. Start wsl.
+```
+wsl
+```
+3. Install the pxt-microbit dependencies.
+```
+npm install
```
Comment (eanders-ms) [inline README.md]: I believe you will need --ignore-scripts here too.
```suggestion
npm install --ignore-scripts
```

File: assets.ts [hunk 2]
```diff
@@ -1,3608 +1,1093 @@
-namespace microcode {
-    let extraImage: SImage = null
-
-    //% shim=TD_NOOP
-    function extraSamples(name: string) {
-        if (name == "clap_lights") extraImage = icondb.sampleClapLights
```
Comment (eanders-ms) [inline assets.ts:6]: Just wondering, why did you need to delete all these assets?

File: clearDataloggerScreen.ts [hunk 3]
```diff
@@ -0,0 +1,105 @@
+namespace microcode {
+    /**
+     * Grant the user the option to clear the contents stored on the data logger.
+     * Inbetween Home and SensorSelect
```
Comment (eanders-ms) [inline clearDataloggerScreen.ts]:
```suggestion
     * In between Home and SensorSelect
```

File: dataRecorder.ts [hunk 4]
```diff
@@ -1,194 +1,394 @@
...
+            )
+
+            this.log()
+        }
+ 
+
+        /**
+         * Schedules the sensors and orders them to .log()
+         */
+        log() {
```
Comment (eanders-ms) [inline dataRecorder.ts]: This function seems to be functionally correct, as far as I can tell, but it has a lot of room for improvement in terms of readability and maintainability. It's very difficult to follow. Consider refactoring each loop into its own function, and inner loop logic into their own functions as well. Local methods would be a nice way to keep the refactored parts private to this operation.

File: dataViewSelect.ts [hunk 5]
```diff
@@ -1,105 +1,122 @@
...
+        constructor(app: App) { 
+            super(app, function () {app.popScene(); app.pushScene(new Home(this.app))})
+        }
+
+        /* override */ startup() {
+            super.startup()
+
+            // Includes the header:
+            this.dataloggerEmpty = true
+            if (datalogger.getNumberOfRows() > 1) {
```
Comment (eanders-ms) [inline dataViewSelect.ts]: Can this change while the DataViewSelect is open? If so, consider making this a getter and query the datalogger for the value. If not, is it possible that might change? If so, this code will break. It might be safer to just query the datalogger.

File: generateGraph.ts [hunk 6]
```diff
@@ -1,157 +1,456 @@
...
+        private windowTopBuffer: number;
+        private windowBotBuffer: number;
+
+        private yScrollOffset: number;
+        private xScrollOffset: number;
+        private xCoordinateScalar: number;
+        private uiState: UI_STATE;
+
+        /** Chunk of 10 data points from the datalogger */
+        private dataRows: string[][];
```
Comment (eanders-ms) [inline generateGraph.ts]: It would be helpful to use the same base identifier for naming `dataRows` and `getNextDataChunk` so that it's clear the latter populates the former. Some possibilitites:
* `dataRows` and `updateDataRows`
* `dataSample` and `getNextDataSample`
* etc.

File: generateGraph.ts [hunk 7]
```diff
@@ -1,157 +1,456 @@
...
+        /** Normalise dataRows for the current screen size: invoked upon UP, DOWN, LEFT, RIGHT */
+        private processedReadings: number[][];
```
Comment (eanders-ms) [inline generateGraph.ts]: The presence of multiple 2d arrays tells me that there's probably a better way to represent this processed data. One way that might make sense is to create a structure to hold multiple processed fields so that they're not split into individual arrays.

File: generateGraph.ts [hunk 8]
```diff
@@ -1,157 +1,456 @@
...
+            this.sensorNames.push(firstSensorName)
+            this.drawSensorStates.push(true)
+            this.sensorMinsAndMaxs.push([sensor.getMinimum(), sensor.getMaximum()])
+            this.numberOfSensors = 1
+            
+            this.lowestSensorMinimum = SENSOR_LOOKUP_TABLE[firstSensorName].getMinimum()
+            this.highestSensorMaximum = SENSOR_LOOKUP_TABLE[firstSensorName].getMaximum()
+
+            // Count until sensor name is repeated:
+            // Go from second sensor onward (3rd row):
```
Comment (eanders-ms) [inline generateGraph.ts]: What is meant by "3rd row"?

File: generateGraph.ts [hunk 9]
```diff
@@ -1,157 +1,456 @@
...
+                    this.processReadings()
+                    this.update() // For fast response to the above changes
+                }
+            )
+
+            control.onEvent(
+                ControllerButtonEvent.Pressed,
+                controller.right.id,
+                () => {
+                    basic.showNumber(6)
```
Comment (eanders-ms) [inline generateGraph.ts]: `basic.showNumber(6)`, is this for debugging?

File: generateGraph.ts [hunk 10]
```diff
@@ -1,157 +1,456 @@
...
+
+        /**
+         * Used to retrieve the next chunk of data
+         * Can be used to intialise this.dataRows
+         * Invoked when this.xScrollOffset changes (Left or Right is pressed)
+         *@returns dataRows where each row contains 4 columns [Sensor Name, Time(ms), Reading, Events]
+         */
+        private getNextDataChunk() {
+            this.dataRows = []
+            const tokens = datalogger.getRows(this.xScrollOffset * 10, (this.xScrollOffset * 10) + 10).split("_");
```
Comment (eanders-ms) [inline generateGraph.ts]: A general guideline: Code should be as self-explanatory as possible, so it can be understood at a glance.
```suggestion
            const startIndex = this.xScrollOffset * 10;
            const rows = datalogger.getRows(startIndex, startIndex + 10);
            const tokens = rows.split("_");

```

File: generateGraph.ts [hunk 11]
```diff
@@ -1,157 +1,456 @@
...
+            this.yScrollOffset = 0
+            this.xScrollOffset = 0
+
+            this.getNextDataChunk()
+            this.sensorNames = []
+            this.drawSensorStates = []
+            this.sensorMinsAndMaxs = []
+
+            // Get the minimum and maximum sensor readings and the number of sensors:
+            const firstSensorName: string = this.dataRows[0][0]
```
Comment (eanders-ms) [inline generateGraph.ts]: Can `this.dataRows` be empty here?

File: generateGraph.ts [hunk 12]
```diff
@@ -1,157 +1,456 @@
...
+            for (let row = 0; row < this.dataRows.length - this.numberOfSensors; row++) {
+                if (this.drawSensorStates[row % this.numberOfSensors]) {
+                    const sensorName: string = this.dataRows[row][0];
+                    const sensor: Sensor = SENSOR_LOOKUP_TABLE[sensorName];
+                    const minimum: number = sensor.getMinimum();
+                    const maximum: number = sensor.getMaximum();
+
+                    // Find the next reading (each sensor can have a different number of measurements)
+                    // So they may not neccessarily be spaced by this.numberOfSensors
+                    const reading1 = this.dataRows[row][2];
```
Comment (eanders-ms) [inline generateGraph.ts]: The meaning of the hard-coded column indices isn't clear. Can you make constants for them? Like:
```ts
    enum SENSOR_COLUMNS {
        NAME = 0,
        VALUE = 2,
        etc...
    }
```

File: generateGraph.ts [hunk 13]
```diff
@@ -1,157 +1,456 @@
...
+
+                screen.print(
+                    this.sensorNames[i],
+                    12,
+                    y + 2,
+                    textColor
+                )
+
+                screen.print(
+                    "Minimum: " + this.sensorMinsAndMaxs[i][0],
```
Comment (eanders-ms) [inline generateGraph.ts]: named constants for the `sensorMinsAndMaxs` second indexer would be helpful for readability.
```suggestion
                    "Minimum: " + this.sensorMinsAndMaxs[i][MIN],

```

File: generateGraph.ts [hunk 14]
```diff
@@ -1,157 +1,456 @@
...
+
+    /**
+     * Takes the datalogger logs and generates a labelled graph.
+     * Each sensor is a unique coloured line, sensor information is detailed below.
+     * 
+     * GRAPH GENERATION IS NOT PERFECT:
+     *      IF LOGGED DATA HAS VARIABLE MEASUREMENTS THERE IS A CHANCE THE GRAPH COULD BE WRONG
+     *      THERE NEED TO BE MORE CHECKS ON EACH ROW OF DATA TO SOLVE THIS.
+     */
+    export class GraphGenerator extends Scene {
```
Comment (eanders-ms) [inline generateGraph.ts:56]: While this class may function properly, the implementation is hard to follow. That makes it hard to maintain over time, especially by other developers. If this code is meant to be long-lived, I suggest reworking it so that it it understandable at a glance. I'd be happy to work with you on this, but is outside the scope of this PR.

**Review by eanders-ms — COMMENTED** [review]
Kier, over all this looks great. I think the implementation is a little rough in a few places, and could use some refactoring to improve code clarity and maintainability. Specifically `genrateGraph.ts` may benefit from another cleanup pass. I'd be happy to work with you on that (outside of this PR).

I didn't fully understand how individual sensors are informed of the requirements of a given visualizer (four columns required for a graph, for example). It might be a good idea to include guidance for:
* Adding a new visualizer, and how to ensure it is compatible with existing sensors.
* Adding a new sensor, and how to ensure it is compatible with existing visualizers.

File: assets.ts (hunk 2 above)
Comment (KierPalin) [reply assets.ts:6]: Tom reccommended I cut excess files and assets as much as possible.
